amm/contracts/stable_pool/lib.rs

```diff
@@ -305,47 +293,19 @@ pub mod stable_pool {
             self.pool.tokens[token_id].into()
         }
 
-        /// Update cached rates if expired.
-        ///
-        /// NOTE:
-        /// If the pool contains a token with rate oracle, this function makes
-        /// a cross-contract call to the `RateProvider` contract if the cached rate is expired.
-        /// This means that the gas cost of the contract methods that use this function may vary,
-        /// depending on the state of the expiration timestamp of the cached rate.
-        fn update_rates(&mut self) {
-            let current_time = self.env().block_timestamp();
-            let mut rate_changed = false;
-            for rate in self.pool.token_rates.iter_mut() {
-                rate_changed |= rate.update_rate(current_time);
-            }
-            if rate_changed {
-                Self::env().emit_event(RatesUpdated {
-                    rates: self.pool.token_rates.clone(),
-                });
-            }
-        }
-
-        /// Get updated rates
-        fn get_updated_rates(&self) -> Vec<TokenRate> {
-            let current_time = self.env().block_timestamp();
-            let mut rates = self.pool.token_rates.clone();
-            rates.iter_mut().for_each(|rate| {
-                rate.update_rate(current_time);
-            });
-            rates
-        }
-
         /// Scaled rates are rates multiplied by precision. They are assumed to fit in u128.
         /// If TOKEN_TARGET_DECIMALS is 18 and RATE_DECIMALS is 12, then rates not exceeding ~340282366 should fit.
         /// That's because if precision <= 10^18 and rate <= 10^12 * 340282366, then rate * precision < 2^128.
-        fn get_scaled_rates(&self, rates: &[TokenRate]) -> Result<Vec<u128>, MathError> {
-            rates
-                .iter()
+        fn get_scaled_rates(&mut self) -> Result<Vec<u128>, MathError> {
+            let current_block_no = self.env().block_number();
+            self.pool
+                .token_rates
+                .iter_mut()
                 .zip(self.pool.precisions.iter())
                 .map(|(rate, &precision)| {
-                    rate.get_rate()
+                    rate.get_rate(current_block_no)
```

Comment:
`rate.get_rate(...)` seems to always use `current_block_no`. Would you consider changing signature to
`get_rate()` (no arguments) and computing `current_block_no` inside?

Arg to `get_rate(...)` is even called `current_block_no` which suggests that only the current block should be passed. But calls to the method may be misleading because `get_rate(block)` suggests that we can put any block number here, but `get_rate(0)` won't give us rates from the genesis block, but from the last updated block.

Reply:
If you're sure that the change is equivalent then OK :+1: